pennylane/beta/_qubit_device.py

```diff
@@ -0,0 +1,632 @@
+    _transpose = staticmethod(np.transpose)
+    _tensordot = staticmethod(np.tensordot)
+    _conj = staticmethod(np.conj)
+    _imag = staticmethod(np.imag)
+    _roll = staticmethod(np.roll)
+    _stack = staticmethod(np.stack)
+    _outer = staticmethod(np.outer)
+    _diag = staticmethod(np.diag)
+
+    @staticmethod
+    def _scatter(indices, array, new_dimensions):
+        new_array = np.zeros(new_dimensions, dtype=array.dtype.type)
+        new_array[indices] = array
+        return new_array
+
+    observables = {"PauliX", "PauliY", "PauliZ", "Hadamard", "Hermitian", "Identity"}
+
+    def __init__(self, wires=1, shots=1000, analytic=True):
+        super().__init__(wires=wires, shots=shots)
+
+        self.analytic = analytic
+        """bool: If ``True``, the device supports exact calculation of expectation
+        values, variances, and probabilities. If ``False``, samples are used
+        to estimate the statistical quantities above."""
+
+        self._samples = None
+        """None or array[int]: stores the samples generated by the device
+        *after* rotation to diagonalize the observables."""
+
+        self._circuit_hash = None
+        """None or int: stores the hash of the circuit from the last execution which
+        can be used by devices in :meth:`apply` for parametric compilation."""
+
+    @classmethod
+    def capabilities(cls):
+
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="qubit",
+            supports_finite_shots=True,
+            supports_tensor_observables=True,
+            returns_probs=True,
+        )
+        return capabilities
+
+    def reset(self):
+        """Reset the backend state.
+
+        After the reset, the backend should be as if it was just constructed.
+        Most importantly the quantum state is reset to its initial value.
+        """
+        self._samples = None
+        self._circuit_hash = None
+
+    def execute(self, circuit, **kwargs):
+        """Execute a queue of quantum operations on the device and then
+        measure the given observables.
+
+        For plugin developers: instead of overwriting this, consider
+        implementing a suitable subset of
+
+        * :meth:`apply`
+
+        * :meth:`~.generate_samples`
+
+        * :meth:`~.probability`
+
+        Additional keyword arguments may be passed to the this method
+        that can be utilised by :meth:`apply`. An example would be passing
+        the ``QNode`` hash that can be used later for parametric compilation.
+
+        Args:
+            circuit (~.CircuitGraph): circuit to execute on the device
+
+        Raises:
+            QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
+
+        Returns:
+            array[float]: measured value(s)
+        """
+        self.check_validity(circuit.operations, circuit.observables)
+
+        self._circuit_hash = circuit.hash
+
```

Comment:
Changes below: Outsource the rest of this function to ``single_execute``, and add the ``batch_execute`` option.